owasp-top10-2017-apps/a7/streaming/README.md

```diff
@@ -0,0 +1,91 @@
+
+The main goal of this app is to discuss how **Cross-Site Scripting** vulnerabilities can be exploited and to encourage developers to send secDevLabs Pull Requests on how they would mitigate these flaws.
+
+## Setup
+
+To start this intentionally **insecure application**, you will need [Docker][Docker Install] and [Docker Compose][Docker Compose Install]. After forking [secDevLabs](https://github.com/globocom/secDevLabs), you must type the following commands to start:
+
+```sh
+cd secDevLabs/owasp-top10-2017-apps/a7/streaming
+```
+
+```sh
+make install
+```
+
+Then simply visit [localhost:10007][App] ! 😆
+
+## Get to know the app
+
+When you access the Streaming application you will be identified as anonymous user and can to watch a streaming on registered users channel and interact with another users (or the master channel) through messages on chat.
+
+## Attack narrative
+
+Now that you know the purpose of this app, what could go wrong? The following section describes how an attacker could identify and eventually find sensitive information about the app or its users. We encourage you to follow these steps and try to reproduce them on your own to better understand the attack vector! 😜
+
+#### Non-sanitization of user input allows for cross-site scripting
+
+After reviewing `buildLiveHTMLMessage(message)` from [`play.component.ts`]((https://github.com/globocom/secDevLabs/blob/master/owasp-top10-2017-apps/a7/streaming/app/frontend/src/app/lives/play/play.component.ts#)) file, it was possible to identify that loaded messages and username are not sanitized and can be executed on a web browser. The following images show this behavior when the following text  is used as an input on these fields:
+
+```
+<b><i>Hi</i></b>
+```
+
+Adding a new message on chat:
+   <img src="images/attack-1.png" align="center"/>
+
+   <img src="images/attack-2.png" align="center"/>
+
+The missing message validation (that will be loaded by another users) allows a malicious user to insert some scripts that will persist in the server and be executed on the victims' browser every time they access the routes that contain these scripts.
+
+### 🔥
+
+An attacker may abuse these flaws by generating a malicious HTML/JS code and sending it to other users. To demonstrate this, the following code example will redirect all users that are watching the channel to another channel.
+
+```html
+<img src=x onError="window.location.href='http://localhost:10007/play/@mr.robot'"/>
+```
+
+This code redirect all users to another page, in this case is to **/play/@mr.robot** route.
+
+When the message is loaded to victim, the browser will interpret the text and will try load image, however the informed image is invalid, then the javascript function ```window.location.href``` will be executed. The following gif show the attacker sending the malicious code to redirect victims (that are watching the **@matthewpets** live) to **/play/@mr.robot** route:
```

Comment:
Would you mind fixing this typo? 🙂 

`When the message is loaded to victim, the browser will interpret the text and will try load image, however the informed image is invalid, then the javascript function ```window.location.href``` will be executed. The following gif show the attacker sending the malicious code to redirect victims (that are watching the **@matthewpets** live) to **/play/@mr.robot** route:`

To:

`When the message is loaded by the victim, the browser will read it and try to load the image, however, the path is invalid. Subsequently, the JavaScript function ```window.location.href``` will be executed.`

`The following gif shows the attacker sending the malicious code to redirect victims (that are watching **@matthewpets** live) to **/play/@mr.robot** route:`